java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java

```diff
@@ -129,18 +135,60 @@ public String valueAsString() {
     .addWithoutParametersMatcher()
     .build();
 
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.tryWithResourcesTrees.clear();
+    this.knownResources.clear();
+    super.scanFile(context);
+  }
+
   @Override
   public void init(MethodTree methodTree, CFG cfg) {
     this.visitedMethodOwnerType = methodTree.symbol().owner().type();
   }
 
   @Override
   public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    collectTryWithResources(syntaxNode);
     final PreStatementVisitor visitor = new PreStatementVisitor(context);
     syntaxNode.accept(visitor);
     return visitor.programState;
   }
 
+  private void collectTryWithResources(Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.TRY_STATEMENT)) {
+      TryStatementTree tryStatementTree = (TryStatementTree) syntaxNode;
+      ListTree<Tree> resourceList = tryStatementTree.resourceList();
+      if (!resourceList.isEmpty() && tryWithResourcesTrees.add(tryStatementTree)) {
+        knownResources.addAll(collectAllResourceTrees(resourceList));
+      }
+    }
+  }
+
+  private static Set<Tree> collectAllResourceTrees(Tree tree) {
```

Comment:
Question: I would typically use a `BaseTreeVisitor` to do something like this, is there any reasons you are using something else?

Reply:
I initially wanted to collect all possible trees, without exception, and with a `BaseTreeVisitor` I would have had to override all the methods to do so. On a second pass, I only collected identifiers, method invocations, and initializers, so the approach with a visitor is indeed easier to get.